lib/Conversion/HandshakeToHW/HandshakeToHW.cpp

```diff
@@ -47,18 +47,45 @@ struct HandshakeLoweringState {
   NameUniquer nameUniquer;
 };
 
-// Wraps a type into an ESI ChannelType type. The inner type is converted to
-// ensure comprehensability by the RTL dialects.
-static Type esiWrapper(Type t) {
-  // Translate index types to something HW understands.
-  if (t.isa<IndexType>())
-    t = IntegerType::get(t.getContext(), 64);
+static Type tupleToStruct(TupleType tuple) {
+  auto *ctx = tuple.getContext();
+  mlir::SmallVector<hw::StructType::FieldInfo, 8> hwfields;
+  for (auto [i, innerType] : llvm::enumerate(tuple))
+    hwfields.push_back(
+        {StringAttr::get(ctx, "field" + std::to_string(i)), innerType});
```

Comment:
Tuples can be nested, so the `innerType` should be converted as well.

Reply:
good catch!